### ISSUE TITLE
Vault SDB always returns the query URL for salt and salt-call. Salt-run executes properly.

### ISSUE THREAD

**Issue by MDPolaris** [saltstack/salt#39163]
### Description of Issue/Question
I setup vault as an external pillar and was trying to use the SDB modules and ran into some strange behavior. I followed the tutorial in the documentation and setup both the vault profile and ext_pillar definition. The external pillar works fine, however when I try to sdb.get sdb://URL, the minion returns only the URL: "sdb://URL" I looked through all the docs on SDB in general and Vault SDB and I couldn't find anything to indicate what the issue could be. I also tried the #salt IRC channel but I had no luck there either.

### Setup
I configured a separate profile for the pillar and sdb in case this was the issue:
```
myvault:
  vault.host: VAULT_IP
  vault.port: 8200
  vault.scheme: http
  vault.token: 


sdb_vault:
  driver: vault
  vault.host: VAULT_IP (Same vault instance as above)
  vault.port: 8200
  vault.scheme: http
  vault.token: 

ext_pillar:
  - vault: myvault path=secret/salt
```

**Vault Values:**
```
$ vault read /secret/salt
Key                     Value
---                     -----
refresh_interval        768h0m0s
rds_pass                VAULT_SECURED_CHANGED
twx_login               RANDOM_PASS_CHANGED

```

### Steps to Reproduce Issue
Using the pillar works:
```
# salt 'minion' pillar.get twx_login
minion:
    RANDOM_PASS_CHANGED

```
Using SDB does not work with either profile:
```
# salt 'minion' sdb.get sdb://sdb_vault/secret/salt?twx_login
minion:
    sdb://sdb_vault/secret/salt?twx_login

# salt 'minion' sdb.get sdb://myvault/secret/salt?twx_login
minion:
    sdb://myvault/secret/salt?twx_login
```

It doesn't seem to matter what I enter, valid or not:
```
# salt 'minion' sdb.get sdb://garbage
minion:
    sdb://garbage

```
salt-run does work with the SDB profile that defines the driver:
```
# salt-run sdb.get sdb://sdb_vault/secret/salt?twx_login
RANDOM_PASS_CHANGED

```
salt-run does not work with the pillar based profile with no driver defined:
```
# salt-run sdb.get sdb://myvault/secret/salt?twx_login
sdb://myvault/secret/salt?twx_login

```

### Versions Report

# salt --versions-report
Salt Version:
           Salt: 2016.11.1

Dependency Versions:
           cffi: 1.6.0
       cherrypy: 3.2.2
       dateutil: 1.5
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.24.0
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: 0.24.0
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.3.1611 Core
        machine: x86_64
        release: 3.10.0-327.10.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core

# salt 'minion' test.versions_report
minion:
    Salt Version:
               Salt: 2016.11.1

    Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: Not Installed
              gitdb: Not Installed
          gitpython: Not Installed
              ioflo: Not Installed
             Jinja2: 2.7.2
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: Not Installed
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: 2.6.1
             pygit2: Not Installed
             Python: 2.7.5 (default, Nov 20 2015, 02:00:19)
       python-gnupg: Not Installed
             PyYAML: 3.10
              PyZMQ: 15.3.0
               RAET: Not Installed
              smmap: Not Installed
            timelib: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.1.4

    System Versions:
               dist: centos 7.2.1511 Core
            machine: x86_64
            release: 3.10.0-327.10.1.el7.x86_64
             system: Linux
            version: CentOS Linux 7.2.1511 Core

**Comment by Ch3LL** [comment]
@MDPolaris as this is not an area i'm an expertise in I looked at the code and it does look like it will return the url in a couple of circumstances. If you look at the code [here](https://github.com/saltstack/salt/blob/v2016.11.2/salt/utils/sdb.py#L13)


As that code shows the following instances will be the cause of returning the url:

```
    if not isinstance(uri, string_types):
        return uri

    if not uri.startswith('sdb://'):
        return uri

    comps = uri.replace('sdb://', '').split('/', 1)

    if len(comps) < 2:
        return uri

    profile = opts.get(comps[0], {})
    if 'driver' not in profile:
        return uri
```

Would you mind adding some logging statements there to find out where its returning uri?

While you do that maybe @techhat has any ideas on this since he is more of the expert in this area.

**Comment by techhat** [comment]
Yes, if SDB has any problems resolving the URI, then it is designed to return it as is.

`salt-run` will not resolve pillars because runners are designed to be a master-side operation and pillars are minion-side only. In order for `salt-run` SDB to work, you must either have the configuration in the `master` config file, or in the case of a local run (`salt-call --local`), the `minion` file.

I'm not sure why the execution module isn't resolving `sdb.get` properly. I'd be interested to see whether `config.get` works though, since it also resolves SDB using the same code as the execution module uses, but without using the execution module itself.

**Comment by MDPolaris** [comment]
Thank you Megan and Joseph,

I added some logging and determined that SDB is exiting at this block:
https://github.com/saltstack/salt/blob/v2016.11.2/salt/utils/sdb.py#L30

This is failing because sdb_vault (the value of comps[0]) is not in the
profile:
Comps: ['sdb_vault', 'secret/salt?new_key']

The list of keys is far too long to include, but here is the snipped of the
sorted keys found by calling sorted(opts.keys()):
'salt_event_pub_hwm', 'schedule', 'sls_list', 'snapper_states'

BTW config.get just returns with no output at all:
# salt 'MINION' config.get sdb://sdb_vault/secret/salt?new_key
MINION:


If I understand this correctly, the SDB config profile is not being sent to
the minions, and that is why this is failing when using salt 'MINION'
sdb.get... The profile does exist for the master when using the runner. Is
this expected behavior? From the docs and example it looks like this should
work when defined on the master only. Thanks.

Mike

**Comment by Ch3LL** [comment]
Ohh okay yeah `config.get` will not be able to pull in master config data unless `pillar_opts` is set to true shown [here](https://docs.saltstack.com/en/latest/ref/configuration/master.html#pillar-opts) BUT pillar_opts is set to False by default for a reason. Because if you are passing in passwords this will be passed over as well. So I believe you have a couple of options:
1. add sdb info to minion config
2. add sdb infor to pillar (not sure if this works)
3. add sdb to master and enable pillar_opts with the understanding that you are passing through passwords.

ping @techhat is my assumption correct here ^ He would need to add the sdb information to hte minion if he wants to run sdb.get? or config.get?

**Comment by JDiPierro** [comment]
Hiya. I'm trying to set up Vault as and SDB source as well and am having quite the headache. The behavior of returning the URI when it can't be parsed seems very strange; why not throw an error?

**Comment by Ch3LL** [comment]
@JDiPierro makes sense to me. Would you mind opening another issue so we can focus on helping @MDPolaris get in a working state in this issue and we can focus on adding more error details in the issue you open. Thanks

**Comment by MDPolaris** [comment]
Hi Megan,

Sorry for the delay. I got tied up on cleaning up some projects before
going out on vacation. Your suggestions make sense and i definitely do not
want to use pillar_opts. Now that I understand how this works better I will
just try to use the runner for SDB. I prefer to keep all of the SDB config
locked away in the master if I can. I will also look into updating the docs
to clarify this as well. I have not contributed before so i need to review
the docs on contributing doc updates. Thanks again. I will let you know if
using the runner does not work for me, but I suspect it will. I will also
investigate the different failure modes as discussed. I need to make sure
that I can properly parse a bad return from SDB, ie: PATH returned vs error
vs real data, etc.Thanks.

Mike

**Comment by Ch3LL** [comment]
Awesome we love it when people contribute. Please let me know if i can help in anyway! I'll keep this open as a documentation issue :)

**Comment by MDPolaris** [comment]
Hi Megan,

I am going to review these pages before starting my contribution, is there
anything else I should look at?

https://docs.saltstack.com/en/latest/topics/development/contributing.html
https://docs.saltstack.com/en/latest/topics/development/conventions/documentation.html#salt-docs

Back on to the SDB module, specifically the Vault one. From what I can
tell, only one single key/value pair can be written with SDB, however Vault
supports multiple key/value pairs at each location, for example:

$ vault read /secret/salt/pillar
Key                     Value
---                     -----
refresh_interval        768h0m0s
second_key              added_value
vault_pillar            test_value

If you are not familiar with Vault, the refresh_interval is a system value,
"second_key" and "vault_pillar" are the keys that I wrote into the system.
When i try this with SDB I can read the individual values:

# salt-run sdb.get sdb://sdb_vault/secret/salt/pillar?second_key
added_value

# salt-run sdb.get sdb://sdb_vault/secret/salt/pillar?vault_pillar
test_value

However you can't write 2 values (I did not expect this to work):

# salt-run sdb.set sdb://sdb_vault/secret/salt/pillar?sdb_1 test1 sdb_2
test2
Exception occurred in runner sdb.set: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/client/mixins.py", line 395,
in _low
    data['return'] = self.functions[fun](*args, **kwargs)
TypeError: set_() takes exactly 2 arguments (4 given)

This would not be a problem, except by design Vault overwrites the existing
values when writing to a location. This means we have to read all the
values from a location and write them back in when making any updates. I
did notice that there is a very large merge for all of the Vault code and
it looks like this may even be fixed:
https://github.com/saltstack/salt/commit/e5ecd8ecf25c0c3c92cc4ad4ea047c5a1d31f5c1

I am still trying to pour through the code there. I can see the new
execution module supports multiple key/value pairs:
https://github.com/saltstack/salt/blob/e5ecd8ecf25c0c3c92cc4ad4ea047c5a1d31f5c1/salt/modules/vault.py#L97

However it looks to me that the SDB system will not gain this functionality:
https://github.com/saltstack/salt/blob/e5ecd8ecf25c0c3c92cc4ad4ea047c5a1d31f5c1/salt/sdb/vault.py#L55

Should I contact the developer directly about these questions? The source
of this code is: https://github.com/carlpett/salt-vault

Thanks again.

Mike

**Comment by truches** [comment]
this weekk i try sdb on salt.  I try many way. When i execute highstate, nothing assign to my value. My question is : can i have a easy example to try it

**Comment by epcim** [comment]
have you set sdb properly in minion conf? 
https://docs.saltstack.com/en/develop/ref/sdb/all/salt.sdb.vault.html#module-salt.sdb.vault

try first with sqlite ..

are you using ext_pillars?

**Comment by truches** [comment]
I try many many hour.  Is it possible to have a step by step example. Every time i try vault sdb sqlite3 work but unable to get result from salt sls

Envoyé de m

Le 28 juin 2017 à 14:37, Petr Michalec <notifications@github.com<mailto:notifications@github.com>> a écrit :


have you set sdb properly in minion conf?
https://docs.saltstack.com/en/develop/ref/sdb/all/salt.sdb.vault.html#module-salt.sdb.vault

try first with sqlite ..

are you using ext_pillars?

**Comment by truches** [comment]
hi, i'am back from my vacancy.  I search the best way to store sensitive data like password outside of .sls and pillar. Do have something with a install procedure.  I check gpg, vault, sdb, consul, hiera .  I am lost.

thank


________________________________
De : Petr Michalec <notifications@github.com>
Envoyé : 28 juin 2017 18:37
À : saltstack/salt
Cc : truches; Comment
Objet : Re: [saltstack/salt] Vault SDB always returns the query URL for salt and salt-call. Salt-run executes properly. (#39163)


have you set sdb properly in minion conf?
https://docs.saltstack.com/en/develop/ref/sdb/all/salt.sdb.vault.html#module-salt.sdb.vault

salt.sdb.vault module - SaltStack Documentation<https://docs.saltstack.com/en/develop/ref/sdb/all/salt.sdb.vault.html#module-salt.sdb.vault>
docs.saltstack.com
This module allows access to Hashicorp Vault using an sdb:// URI. Base configuration instructions are documented in the execution module docs. Below are noted extra ...


try first with sqlite ..

are you using ext_pillars?

**Comment by golmaal** [comment]
I was trying consul and having the exact issue as Mike. Glad I searched 😄 

Thanks for posting suggestions @Ch3LL 

If I copy my sdb configuration to minion.d, salt-call sdb.get sdb://myconsul/salt-private/node-01/tomcat starts to work... 
myconsul:
  driver: consul 

Question for you @Ch3LL - Is there a downside to making the SDB configuration available to all minions? I'm trying to automate creating a tomcat user for manager-script role and save its password in tomcat-users.xml file and also in the pillar specific to that node. My purpose is to be able to deploy war files using tomcat execution module with pillar data in a state file. I would like to use sdb.get_or_set_hash function when deploying tomcat and sdb.get when deploying war files.

**Comment by Ch3LL** [comment]
How are you making the configuration available to minions?

If you are setting it in the master config file by enabling pillar_opts you could be sending sensitive information that you don't want the minion to have access to. So it depends on how you set it up

**Comment by golmaal** [comment]
As I was trying etcd and consul (really to decide which one was easier for redistribution purpose), I see both allow use of private/%(minion_id)s
This allows the correct pillar information to be populated for each minion. 
Earlier when I wrote the comment above I was thinking to call sdb.get_or_set_hash from states which would have required me to let minion config know about the sdb profile through driver setting. 
Just to be explicit, I'm not setting pillar_opts.
Here is all of my configuration - 

$ cat /etc/salt/master.d/consul.conf
consul_config:
  consul.host: 127.0.0.1
  consul.port: 8500

$ cat /etc/salt/master.d/sdb.conf
myconsul:
  driver: consul

$ cat /etc/salt/master.d/pillar.conf
ext_pillar:
  - consul: consul_config root=salt-private/%(minion_id)s

If I copy /etc/salt/master.d/sdb.conf to /etc/salt/minion.d/ folder as well, then I can use sdb.get or sdb.set using salt-call as well. That was the comment I made originally in this thread.

**Comment by Ch3LL** [comment]
Okay thanks for clarifying. To answer your question: Is there a downside to making the SDB configuration available to all minions? I would not see a downside other then you will need to manage this across all minions, which of course you could do in a state.

but maybe @techhat or @gtmanfred know of any disadvantages?

**Comment by gtmanfred** [comment]
If you upgrade to 2017.7, the vault configuration states on the salt master, and only the sdb configuration goes on the minion, and the only sdb configuration is

```
myvault:
  driver: vault
```

The vault configuration all goes on the master https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.vault.html

And The minion has to make a request to the master for a token that has an expiration time, and then can use that token to request an entry in vault.

**Comment by golmaal** [comment]
Thanks Megan @Ch3LL and Daniel @gtmanfred 
I'll take another look at Vault and try to understand it better. 
I was attracted to ext_pillar because of its capability to populate pillar data appropriately when configured with  %(minion_id)s in the root.

I restricted our use of sdb to runners because I thought if it became available to minions, any minion could see data through sdb.get not intended for it. E.g. let's say root=salt-private/%(minion_id)s/tomcat-manager
If sdb.get can be executed on any of the minion, then any minion could supply the name of other minion and retrieve tomcat-manager configuration explicitly...
Do I understand this behavior correctly?

**Comment by gtmanfred** [comment]
What you can do in vault, is you make a bunch of different roles.

And then you assign that minions role in the vault permissions

```
policies:
  - saltstack/minion/{minion}
```

Then the token that is generated is only allowed to view files from that policy.  The master generates the token, and it knows which minion is requesting access, so the minion cannot fudge another minions token.

```
policies

    Policies that are assigned to minions when requesting a token. These can either be static, eg saltstack/minions, or templated, eg saltstack/minion/{minion}. {minion} is shorthand for grains[id]. Grains are also available, for example like this: my-policies/{grains[os]}

    If a template contains a grain which evaluates to a list, it will be expanded into multiple policies. For example, given the template saltstack/by-role/{grains[roles]}, and a minion having these grains:

    The minion will have the policies saltstack/by-role/web and saltstack/by-role/database. Note however that list members which do not have simple string representations, such as dictionaries or objects, do not work and will throw an exception. Strings and numbers are examples of types which work well.

    Optional. If policies is not configured, saltstack/minions and saltstack/{minion} are used as defaults.

```

Daniel

**Comment by golmaal** [comment]
Thanks Daniel. I'll give it a try. This is useful information!
